File: web_frontend/frontend/src/api/agentPackages.ts
import { requestBlob, requestEvent, requestJson, withQuery } from './http'

interface RecentAgentSessionsResponse {
  sessions: any[]
}

interface AgentPackageConfigurationResponse {
  package: any
}

export type AgentToolApproval = 'inherit' | 'allow' | 'ask' | 'deny'
export type AgentToolPolicyMode = 'strict' | 'allow_below_high' | 'allow_all' | 'custom'

export interface AgentToolSettingView {
  tool_id: string
  name: string
  description: string
  base_description: string
  description_overridden: boolean
  source: string
  risk_level: 'low' | 'medium' | 'high'
  permission_scope: string
  permission_tags: string[]
  max_model_chars: number
  max_model_chars_overridden: boolean
  base_concurrent: boolean
  concurrent: boolean
  concurrent_overridden: boolean
}

export interface AgentToolSettingsView {
  policy: {
    mode: AgentToolPolicyMode
    low?: string
    medium?: string
    high?: string
    tool_overrides: Record<string, { approval?: AgentToolApproval; risk_level?: string | null }>
  }
  default_max_model_chars: number
  tools: AgentToolSettingView[]
}

interface AgentToolSettingsResponse {
  tool_settings: AgentToolSettingsView
}

export interface AgentPackageResourceDescriptorView {
  resource_id: string
  description: string
  required: boolean
  configured: boolean
  value?: unknown
  secret_fields: string[]
  used_by: string[]
  sandbox_access_expectation: string
  value_schema: Record<string, unknown>
  key_available: boolean
}

export interface AgentPackageResourcesResponse {
  package_id: string
  key_available: boolean
  resources: AgentPackageResourceDescriptorView[]
  migration: { status: string; migrated?: boolean; reason?: string }
}

export const agentPackagesApi = {
  list: () => requestEvent('/api/agent-packages'),
  select: (packageId: string, purpose?: 'run' | 'evolution') =>
    requestEvent('/api/agent-packages/select', {
      method: 'POST',
      body: JSON.stringify({ package_id: packageId, purpose }),
    }),
  instances: () => requestEvent('/api/agent-packages/instances'),
  recentSessions: (limit = 5) =>
    requestJson<RecentAgentSessionsResponse>(withQuery('/api/agent-packages/recent-sessions', { limit })),
  initialize: (packageId: string) =>
    requestEvent(`/api/agent-packages/${encodeURIComponent(packageId)}/initialize`, { method: 'POST' }),
  shutdown: (packageId: string) =>
    requestEvent(`/api/agent-packages/${encodeURIComponent(packageId)}/shutdown`, { method: 'POST' }),
  delete: (packageId: string) =>
    requestEvent(`/api/agent-packages/${encodeURIComponent(packageId)}`, { method: 'DELETE' }),
  exportArchive: (packageId: string) =>
    requestBlob(`/api/agent-packages/${encodeURIComponent(packageId)}/export`),
  toolSettings: (packageId: string) =>
    requestJson<AgentToolSettingsResponse>(
      `/api/agent-packages/${encodeURIComponent(packageId)}/tool-settings`,
    ),
  updateToolPolicy: (packageId: string, mode: AgentToolPolicyMode) =>
    requestJson<AgentToolSettingsResponse>(
      `/api/agent-packages/${encodeURIComponent(packageId)}/tool-settings/policy`,
      {
        method: 'PATCH',
        body: JSON.stringify({ policy: { mode } }),
      },
    ),
  updateToolSettings: (
    packageId: string,
    toolId: string,
    payload: {
      description: string
      max_model_chars: number
      approval: AgentToolApproval
      concurrent: boolean
    },
  ) =>
    requestJson<AgentToolSettingsResponse>(
      `/api/agent-packages/${encodeURIComponent(packageId)}/tool-settings/${encodeURIComponent(toolId)}`,
      {
        method: 'PATCH',
        body: JSON.stringify(payload),
      },
    ),
  resetToolSettings: (packageId: string, toolId: string) =>
    requestJson<AgentToolSettingsResponse>(
      `/api/agent-packages/${encodeURIComponent(packageId)}/tool-settings/${encodeURIComponent(toolId)}`,
      { method: 'DELETE' },
    ),
  updateContextConfig: (packageId: string, config: Record<string, unknown>) =>
    requestJson<AgentPackageConfigurationResponse>(
      `/api/agent-packages/${encodeURIComponent(packageId)}/context-config`,
      {
        method: 'PATCH',
        body: JSON.stringify({ config }),
      },
    ),
  updateSchedulerConfig: (packageId: string, config: Record<string, unknown>) =>
    requestJson<AgentPackageConfigurationResponse>(
      `/api/agent-packages/${encodeURIComponent(packageId)}/scheduler-config`,
      {
        method: 'PATCH',
        body: JSON.stringify({ config }),
      },
    ),
  updateModelOverrides: (
    packageId: string,
    bindings: Record<string, { temperature: number | null; max_output_tokens: number | null }>,
    toolBindings: Record<string, { temperature: number | null; max_output_tokens: number | null }>,
  ) =>
    requestJson<AgentPackageConfigurationResponse>(
      `/api/agent-packages/${encodeURIComponent(packageId)}/model-overrides`,
      {
        method: 'PATCH',
        body: JSON.stringify({ bindings, tool_bindings: toolBindings }),
      },
    ),
  resources: (packageId: string) =>
    requestJson<AgentPackageResourcesResponse>(`/api/agent-packages/${encodeURIComponent(packageId)}/resources`),
  putResource: (packageId: string, resourceId: string, value: unknown) =>
    requestJson(`/api/agent-packages/${encodeURIComponent(packageId)}/resources/${encodeURIComponent(resourceId)}`, {
      method: 'PUT',
      body: JSON.stringify({ value }),
    }),
  deleteResource: (packageId: string, resourceId: string) =>
    requestJson(`/api/agent-packages/${encodeURIComponent(packageId)}/resources/${encodeURIComponent(resourceId)}`, {
      method: 'DELETE',
    }),
  sessions: (packageId: string) => requestEvent(`/api/agent-packages/${encodeURIComponent(packageId)}/sessions`),
  session: (packageId: string, sessionId: string) =>
    requestEvent(
      `/api/agent-packages/${encodeURIComponent(packageId)}/sessions/${encodeURIComponent(sessionId)}`
    ),
  deleteSession: (packageId: string, sessionId: string) =>
    requestEvent(
      `/api/agent-packages/${encodeURIComponent(packageId)}/sessions/${encodeURIComponent(sessionId)}`,
      { method: 'DELETE' }
    ),
}

File: web_frontend/backend/routes/agent_packages.py
from __future__ import annotations

import logging
from pathlib import Path
from typing import Any
from urllib.parse import quote

from fastapi import APIRouter, HTTPException, Query
from fastapi.responses import FileResponse
from starlette.background import BackgroundTask

from agent_factory.factory_graph.frontend_bridge.agent_package_runtime import AgentPackageRuntimeManager
from web_frontend.backend.runtime_bridge import RuntimeBridge
from web_frontend.backend.routes.utils import send_and_wait

logger = logging.getLogger(__name__)


def create_agent_package_router(runtime_bridge: RuntimeBridge, logger: logging.Logger) -> APIRouter:
    router = APIRouter(prefix="/api/agent-packages")

    @router.get("")
    async def list_agent_packages():
        event = await send_and_wait(runtime_bridge, "list_agent_packages", {}, {"agent_packages_listed"})
        return {"event": event}

    @router.post("/select")
    async def select_agent_package(payload: dict[str, Any]):
        event = await send_and_wait(runtime_bridge, "select_agent_package", payload, {"agent_package_selected"})
        return {"event": event}

    @router.get("/instances")
    async def list_agent_package_instances():
        event = await send_and_wait(
            runtime_bridge,
            "list_agent_package_instances",
            {},
            {"agent_package_instances_listed"},
        )
        return {"event": event}

    @router.get("/recent-sessions")
    def list_recent_agent_package_sessions(limit: int = Query(default=5, ge=1, le=20)):
        runtime = AgentPackageRuntimeManager()
        return {"sessions": runtime.list_recent_sessions(limit=limit)}

    @router.post("/{package_id}/initialize")
    async def initialize_agent_package(package_id: str):
        event = await send_and_wait(
            runtime_bridge,
            "initialize_agent_package",
            {"package_id": package_id},
            {"agent_package_instance_updated"},
            event_filter=lambda item: _is_initialize_status_event(item, package_id=package_id),
        )
        return {"event": event}

    @router.post("/{package_id}/shutdown")
    async def shutdown_agent_package_instance(package_id: str):
        event = await send_and_wait(
            runtime_bridge,
            "shutdown_agent_package_instance",
            {"package_id": package_id},
            {"agent_package_instance_updated"},
        )
        return {"event": event}

    @router.delete("/{package_id}")
    async def delete_agent_package(package_id: str):
        event = await send_and_wait(
            runtime_bridge,
            "delete_agent_package",
            {"package_id": package_id},
            {"agent_package_deleted"},
        )
        return {"event": event}

    @router.get("/{package_id}/export")
    def export_agent_package(package_id: str):
        try:
            archive_path = AgentPackageRuntimeManager().export_package_archive(package_id)
        except FileNotFoundError as exc:
            raise HTTPException(status_code=404, detail=str(exc)) from exc
        except ValueError as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
        filename = f"{package_id}.zip"
        quoted_filename = quote(filename)
        return FileResponse(
            archive_path,
            media_type="application/zip",
            filename=filename,
            headers={"Content-Disposition": f"attachment; filename*=UTF-8''{quoted_filename}"},
            background=BackgroundTask(unlink_file, archive_path),
        )

    @router.get("/{package_id}/tool-settings")
    def get_agent_package_tool_settings(package_id: str):
        try:
            summary = AgentPackageRuntimeManager().extension_config_summary(package_id)
        except FileNotFoundError as exc:
            raise HTTPException(status_code=404, detail=str(exc)) from exc
        except (ValueError, TypeError) as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
        return {"tool_settings": summary.get("tool_permissions", {})}

    @router.patch("/{package_id}/tool-settings/policy")
    def update_agent_package_tool_policy(package_id: str, payload: dict[str, Any]):
        try:
            runtime = AgentPackageRuntimeManager()
            current = runtime.extension_config_summary(package_id)
            current_policy = current.get("tool_permissions", {}).get("policy", {})
            requested_policy = payload.get("policy", payload)
            if not isinstance(requested_policy, dict):
                raise ValueError("tool policy must be an object")
            if requested_policy.get("mode") == "allow_all":
                current_overrides = current_policy.get("tool_overrides", {})
                current_policy = {
                    **current_policy,
                    "tool_overrides": {
                        tool_id: {**override, "approval": "inherit"}
                        for tool_id, override in current_overrides.items()
                        if isinstance(override, dict) and override.get("risk_level") is not None
                    },
                }
            summary = runtime.extensions_manage(
                package_id,
                "update_tool_permissions",
                {"policy": {**current_policy, **requested_policy}},
            )
        except FileNotFoundError as exc:
            raise HTTPException(status_code=404, detail=str(exc)) from exc
        except (ValueError, TypeError) as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
        return {"tool_settings": summary.get("tool_permissions", {})}

    @router.patch("/{package_id}/tool-settings/{tool_id}")
    def update_agent_package_tool_settings(package_id: str, tool_id: str, payload: dict[str, Any]):
        try:
            summary = AgentPackageRuntimeManager().extensions_manage(
                package_id,
                "update_tool_configuration",
                {**payload, "tool_id": tool_id},
            )
        except FileNotFoundError as exc:
            raise HTTPException(status_code=404, detail=str(exc)) from exc
        except (ValueError, TypeError) as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
        return {"tool_settings": summary.get("tool_permissions", {})}

    @router.delete("/{package_id}/tool-settings/{tool_id}")
    def reset_agent_package_tool_settings(package_id: str, tool_id: str):
        try:
            summary = AgentPackageRuntimeManager().extensions_manage(
                package_id,
                "reset_tool_configuration",
                {"tool_id": tool_id},
            )
        except FileNotFoundError as exc:
            raise HTTPException(status_code=404, detail=str(exc)) from exc
        except (ValueError, TypeError) as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
        return {"tool_settings": summary.get("tool_permissions", {})}

    @router.patch("/{package_id}/context-config")
    def update_agent_package_context_config(package_id: str, payload: dict[str, Any]):
        config = payload.get("config")
        if not isinstance(config, dict):
            raise HTTPException(status_code=400, detail="context config is required")
        try:
            summary = AgentPackageRuntimeManager().update_context_config(
                package_id,
                config=config,
            )
        except FileNotFoundError as exc:
            raise HTTPException(status_code=404, detail=str(exc)) from exc
        except (ValueError, TypeError) as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
        return {"package": summary}

    @router.patch("/{package_id}/scheduler-config")
    def update_agent_package_scheduler_config(package_id: str, payload: dict[str, Any]):
        config = payload.get("config")
        if not isinstance(config, dict):
            raise HTTPException(status_code=400, detail="scheduler config is required")
        try:
            summary = AgentPackageRuntimeManager().update_scheduler_config(
                package_id,
                config=config,
            )
        except FileNotFoundError as exc:
            raise HTTPException(status_code=404, detail=str(exc)) from exc
        except (ValueError, TypeError) as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
        return {"package": summary}

    @router.patch("/{package_id}/model-overrides")
    def update_agent_package_model_overrides(package_id: str, payload: dict[str, Any]):
        bindings = payload.get("bindings")
        tool_bindings = payload.get("tool_bindings")
        if not isinstance(bindings, dict) or not isinstance(tool_bindings, dict):
            raise HTTPException(status_code=400, detail="model and model tool overrides are required")
        try:
            summary = AgentPackageRuntimeManager().update_model_overrides(
                package_id,
                bindings=bindings,
                tool_bindings=tool_bindings,
            )
        except FileNotFoundError as exc:
            raise HTTPException(status_code=404, detail=str(exc)) from exc
        except (ValueError, TypeError) as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
        return {"package": summary}

    @router.get("/{package_id}/resources")
    def get_agent_package_resources(package_id: str):
        try:
            return AgentPackageRuntimeManager().resource_status(package_id, include_values=True)
        except FileNotFoundError as exc:
            raise HTTPException(status_code=404, detail=str(exc)) from exc
        except ValueError as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc

    @router.put("/{package_id}/resources/{resource_id}")
    def put_agent_package_resource(package_id: str, resource_id: str, payload: dict[str, Any]):
        if "value" not in payload:
            raise HTTPException(status_code=400, detail="resource value is required")
        try:
            return AgentPackageRuntimeManager().put_resource(package_id, resource_id, payload["value"])
        except FileNotFoundError as exc:
            raise HTTPException(status_code=404, detail=str(exc)) from exc
        except ValueError as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
        except Exception as exc:
            raise HTTPException(status_code=409, detail=str(exc)) from exc

    @router.delete("/{package_id}/resources/{resource_id}")
    def delete_agent_package_resource(package_id: str, resource_id: str):
        try:
            return AgentPackageRuntimeManager().delete_resource(package_id, resource_id)
        except FileNotFoundError as exc:
            raise HTTPException(status_code=404, detail=str(exc)) from exc

    @router.get("/{package_id}/sessions")
    async def list_agent_package_sessions(package_id: str):
        event = await send_and_wait(
            runtime_bridge,
            "list_agent_package_sessions",
            {"package_id": package_id},
            {"agent_package_sessions_listed"},
        )
        return {"event": event}

    @router.get("/{package_id}/sessions/{session_id}")
    async def load_agent_package_session(package_id: str, session_id: str):
        event = await send_and_wait(
            runtime_bridge,
            "load_agent_package_session",
            {"package_id": package_id, "session_id": session_id},
            {"agent_package_session_loaded"},
        )
        return {"event": event}

    @router.delete("/{package_id}/sessions/{session_id}")
    async def delete_agent_package_session(package_id: str, session_id: str):
        event = await send_and_wait(
            runtime_bridge,
            "delete_agent_package_session",
            {"package_id": package_id, "session_id": session_id},
            {"agent_package_session_deleted"},
        )
        return {"event": event}

    return router


def _is_initialize_status_event(item: dict[str, Any], *, package_id: str) -> bool:
    payload = item.get("payload") if isinstance(item.get("payload"), dict) else {}
    return str(payload.get("package_id") or "").strip() == package_id


def unlink_file(path: Path) -> None:
    try:
        path.unlink(missing_ok=True)
    except OSError:
        logger.warning("Failed to remove temporary file: %s", path)
